### DISCUSSION TITLE
Issues using a syntax tree with stringify

### DISCUSSION THREAD

**Discussion by cyberwombat** [unifiedjs/unified#107] [Q&A] [answered]
I am wanting to create two outputs (doc and toc). So I am approaching this by creating a tree that I then pass to two different processors. However my processor is not accepting my tree:
```
import unified from'unified'
import markdown from'remark-parse'
import remark2rehype from 'remark-rehype'
import html from 'rehype-stringify'

// This works
const d = unified().use(markdown).use(remark2rehype).use(html).process(fileContents)

// So extract tree
const tree = unified().use(markdown).parse(fileContents)
// Pass tree to processor - does not work - chokes on unknown type in the tree
const doc = unified().use(remark2rehype).use(html).stringify(tree) 
```
I am using this [md](https://github.com/vercel/next.js/blob/canary/examples/blog-starter-typescript/_posts/preview.md) as sample

What is the proper way to split above?

**Comment by ChristianMurphy** [comment]
> So I am approaching this by creating a tree that I then pass to two different processors

Can you expand on this?
Are you trying to write multiple files? Or run multiple plugins?

> However my processor is not accepting my tree

What do you mean? could you share a codesandbox with an example? (this is a good starting point https://codesandbox.io/s/remark-rehype-debug-4cz8v)

**Comment by cyberwombat** [comment]
Ah thanks for sandbox link - https://codesandbox.io/s/remark-rehype-debug-forked-etych

So I am wanting to have a table of contents that is separate - I already have all the code to render a toc from a tree. So what I am wanting to do is to 1) create a tree then 2) using that tree create the main html and also pass that tree to my toc builder. 

In the sandbox commented out at bottom I have a working example. How can I take that example and end up with the syntax tree AND the final html in a way that minimizes processing.

**Comment by ChristianMurphy** [comment, marked as answer]
A couple problems:
1. unified `stringify` doesn't run transforms, `remark-rehype` is a transform, meaning in your example it is trying to turn a markdown tree into HTML text without having converted to an HTML tree. The overall flow is documented at https://github.com/unifiedjs/unified#description, to make this work `run` is needed to apply the transform.
2. `stringify` is not a promise, calling `.then` will cause an error

see: https://codesandbox.io/s/remark-rehype-debug-forked-4fybi?file=/src/index.js for an example of how this comes together

**Comment by cyberwombat** [comment]
Ah thank you. All trees are not alike :)
